Upgrade Jest from 27 to 29.5.0

jsnajdr opened the pull request:
This is an upgrade of Jest from v27 to v29, two major versions up. There's been a lot of breaking changes, so it's a big and difficult PR.

At this moment, web unit tests are all green, mobile tests still need some fixing.

One thing that's broken in mobile tests are Jest helpers imported from `react-native-reanimated`. Gutenberg uses a [fork of `react-native-reanimated`](https://github.com/wordpress-mobile/react-native-reanimated). For the Jest update, we need to incorporate changes from this PR: https://github.com/software-mansion/react-native-reanimated/pull/3559. @geriux would it be possible to update the fork? Currently the fork is based on upstream version 2.9.1 (Jul 2022), the Jest helper issue (`expect.extend`) has been fixed in version [2.13.0](https://github.com/software-mansion/react-native-reanimated/releases/tag/2.13.0) (Nov 2022) and the latest version is 2.14.4 (Jan 2022, two weeks ago)

jsnajdr committed:
Upgrade Jest and friends to 29.5.0

jsnajdr committed:
Update lockfile

jsnajdr committed:
Migrate the testURL option

jsnajdr committed:
Custom resolver to solve ESM+browser modules

jsnajdr committed:
Use new API to enable legacy fake timers

jsnajdr committed:
isURL tests: remove controversial IPv4-like URLs where the spec recently changed

jsnajdr committed:
Fix mobile unit tests

geriux commented:
> @geriux would it be possible to update the fork? Currently the fork is based on upstream version 2.9.1 (Jul 2022), the Jest helper issue (expect.extend) has been fixed in version [2.13.0](https://github.com/software-mansion/react-native-reanimated/releases/tag/2.13.0) (Nov 2022) and the latest version is 2.14.4 (Jan 2022, two weeks ago)

I think what we could do is cherry-pick that [change](https://github.com/software-mansion/react-native-reanimated/pull/3559/files) into the fork, that would be quicker than updating to a newer version as that requires more work and a lot of testing.

jsnajdr commented:
> I think what we could do is cherry-pick that change into the fork

That would be awesome. The issue is that Jest's `expect` package used to be CommonJS, and `require('extend')` used to return the actual `extend` object, but now it's an ESM module and you need to use `require('extend').default` instead.

geriux commented:
> That would be awesome. The issue is that Jest's `expect` package used to be CommonJS, and `require('extend')` used to return the actual `extend` object, but now it's an ESM module and you need to use `require('extend').default` instead.

Cool! I'll find some time to cherry-pick that and bump the Reanimated version, I'll let you know once it is ready 👍

geriux commented:
Hey @jsnajdr 👋  I'm working on a [draft PR](https://github.com/WordPress/gutenberg/pull/47574) that will update the Reanimated version. In the meantime, you could try pointing to [this version](https://github.com/WordPress/gutenberg/pull/47574/files#diff-1f256578dd9ab8dda6c3913a55a588312129f3404bb0cdf0b1ab52f0a6937119R69) and see if this issue is resolved. Thanks!

geriux commented:
Hey there 👋 The Reanimated update is now merged in `trunk`.

jsnajdr commented:
> The Reanimated update is now merged in trunk.

Thanks for the update @geriux, it really fixed a lot of failures in this branch.

Turns out we need to cherry-pick one more patch for `react-native-reanimated`. There is the `react-native-reanimated/lib/index.js` file that contains an "export namespace as default" statement:
```js
export * as default from './Animated';
```
This syntax is however not supported without an extra Babel plugin, so we need to rewrite it to:
```js
import * as Animated from './Animated';
export default Animated;
```
In the upstream `react-native-reanimated` library this was done in https://github.com/software-mansion/react-native-reanimated/pull/3437 (look just at the `src/index.ts` change), and the change is included in recent versions like 2.14.4.

jsnajdr committed:
jest-environment-puppeteer: fix ESM import of jest-environment-node

jsnajdr committed:
Add Babel config for e2e-tests that babel-jest can find

jsnajdr committed:
Remove legacy timers option (real is the default)

jsnajdr committed:
Mobile tests: fix ESM import for jest-environment-jsdom

jsnajdr commented:
All CI checks are green now :+1: Ready for review and merging.

geriux commented:
> Turns out we need to cherry-pick one more patch for `react-native-reanimated`. There is the `react-native-reanimated/lib/index.js` file that contains an "export namespace as default" statement:

Hey there @jsnajdr 👋 sorry for the delay, is this still needed? I see all checks are passing 🟢. Let us know so we can help. Thank you!

jsnajdr commented:
> sorry for the delay, is this still needed? I see all checks are passing

I would be still nice to have, as newer upstream versions have that patch. But it's no longer necessary. For this branch I solved it by adding an explicit Babel config file for `packages/e2e-tests`: https://github.com/WordPress/gutenberg/pull/47388/commits/535303795e317fb84d3db6601b675f2cea8dafbd

The `babel-preset-default` adds the `export-default-from` transform, needed to support this syntax, and also the `reanimated` Babel plugin which transforms the Reanimated worklets and is needed for animations to work.

geriux commented:
> I would be still nice to have, as newer upstream versions have that patch.

I see, we'll plan to bring that then into a patch release of our fork 👍 

>For this branch I solved it by adding an explicit Babel config file for packages/e2e-tests: https://github.com/WordPress/gutenberg/commit/535303795e317fb84d3db6601b675f2cea8dafbd

Awesome, glad to know that's not blocking this PR.

tyxla commented on packages/compose/src/higher-order/with-global-events/test/index.js:
```diff
@@ -22,7 +22,10 @@ jest.mock( '../listener', () => {
 			super( ...arguments );
 
 			this.constructor._instance = this;
+			this._spy();
+		}
 
+		_spy() {
```
I'm curious if this change makes any difference.

jsnajdr replied:
Without this trick, Jest would hoist the `jest.spyOn` calls above the `super()` call, causing an error. For the same reasons I had to wrap `jest.spyOn` calls in IIFE in other files.

It is a Jest bug, which I reported in https://github.com/facebook/jest/pull/13952 and which I'll try to fix. Then we can revert this part of the patch again.

tyxla commented on packages/jest-preset-default/jest-preset.js:
```diff
@@ -24,7 +24,6 @@ module.exports = {
 		'**/?(*.)test.[jt]s?(x)',
 	],
 	testPathIgnorePatterns: [ '/node_modules/', '<rootDir>/vendor/' ],
-	timers: 'real',
```
Nice catch 👍 Seems like we missed that when we defaulted to real timers.

gziolo replied:
We should also update the README to reflect it:

<img width="730" alt="Screenshot 2023-03-08 at 12 58 44" src="https://user-images.githubusercontent.com/699132/223707730-ee9274ae-d1fc-49b4-814d-04c2a980ed44.png">

tyxla commented on packages/scripts/scripts/test-e2e.js:
```diff
@@ -78,4 +75,4 @@ Object.entries( configsMapping ).forEach( ( [ envKey, argName ] ) => {
 
 const cleanUpPrefixes = [ '--puppeteer-', '--wordpress-' ];
 
-jest.run( [ ...config, ...runInBand, ...getArgsFromCLI( cleanUpPrefixes ) ] );
```
The `--runInBand` change seems to deserve its own PR, doesn't it? Or is it for some reason strictly related to this one?

gziolo replied:
Does it mean you can safely run e2e tests in parallel now? I'm curious about what has changed in Jest about that. The challenge we had at the beginning was that some tests could update the shared database in a way that gives false negatives when running assertions.

jsnajdr replied:
Oh, it seems that I misunderstood the original code. I thought that
```js
const runInBand = ! hasRunInBand ? [ '--runInBand' ] : [];
```
is a bug, but now I realize it's there to always enforce `--runInBand`, even if it wasn't passed on command line?

I'll fix that, and submit these `test:performance:debug` changes as a standalone PR.

gziolo replied:
Yes, that was the intent but somehow tests pass on CI so I’d like to know how. Maybe it’s passed explicitly there elsewhere but I can’t locate it.

tyxla commented on packages/jest-preset-default/package.json:
```diff
@@ -31,7 +31,7 @@
 	"dependencies": {
 		"@wordpress/jest-console": "file:../jest-console",
-		"babel-jest": "^27.4.5"
+		"babel-jest": "^29.3.1"
```
This is worth a CHANGELOG entry.

tyxla commented on packages/jest-console/package.json:
```diff
@@ -33,7 +33,7 @@
 	"dependencies": {
 		"@babel/runtime": "^7.16.0",
-		"jest-matcher-utils": "^27.4.2"
+		"jest-matcher-utils": "^29.3.1"
```
This is worth a CHANGELOG entry.

tyxla commented on packages/react-native-editor/src/test/block-parser-code.test.js:
```diff
@@ -1,8 +1,3 @@
-/**
- * Internal dependencies
- */
-import '../globals';
```
I suppose those imports are unnecessary because of:

https://github.com/WordPress/gutenberg/blob/705ab44f2c2797d75740533a85030c194b4fb040/packages/react-native-editor/src/index.js#L10

jsnajdr replied:
Yes, the globals are already imported elsewhere.

tyxla commented on packages/report-flaky-tests/package.json:
```diff
@@ -28,7 +28,7 @@
 		"@actions/core": "^1.8.0",
 		"@actions/github": "^5.0.1",
-		"jest-message-util": "^28.0.2"
+		"jest-message-util": "^29.3.1"
```
This package's CHANGELOG is quite empty, but we still could include an entry for this change there.

tyxla commented on packages/scripts/package.json:
```diff
@@ -61,9 +61,10 @@
 		"fast-glob": "^3.2.7",
 		"filenamify": "^4.2.0",
-		"jest": "^27.4.5",
+		"jest": "^29.3.1",
```
Those changes are worth one (or even multiple) CHANGELOG entries. This package is used broadly externally, so we should be as thorough as possible when introducing major changes like this.

tyxla commented on packages/e2e-tests/package.json:
```diff
@@ -31,8 +31,8 @@
-		"jest-message-util": "^27.4.2",
-		"jest-snapshot": "^27.4.5",
+		"jest-message-util": "^29.3.1",
+		"jest-snapshot": "^29.3.1",
```
We could add a CHANGELOG entry here.

tyxla commented on packages/e2e-tests/babel.config.js:
```diff
@@ -0,0 +1,7 @@
+module.exports = ( api ) => {
+	api.cache( true );
```
Nice 👍 

I can't think of a scenario where we'd actually want to cache the config per environment with `api.cache.using( () => process.env.NODE_ENV )`.

I wonder if enabling this config cache was motivated by something and whether this change affects the performance of the e2e tests.

gziolo replied:
It's for the test suites' source code. Isn't it always equal to `test`?

jsnajdr replied:
I admin I didn't think about proper caching at all when writing this 🙂 I just copied a statement that every other Babel config in the repo also uses.

tyxla replied:
> It's for the test suites' source code. Isn't it always equal to test?

Yup, that's what I essentially was saying with:

> I can't think of a scenario where we'd actually want to cache the config per environment with `api.cache.using( () => process.env.NODE_ENV )`.

tyxla commented on package.json:
```diff
@@ -195,11 +195,12 @@
 		"husky": "7.0.0",
 		"inquirer": "7.1.0",
-		"jest": "27.4.5",
+		"jest": "29.3.1",
```
Should we also bump some peer deps, like here:

https://github.com/WordPress/gutenberg/commit/caab0febf95760b6e4de61cc7663d0d7fa5c149d

I've noticed that being very flexible with dependencies can cause issues with more complex upgrades (upgrading `@wordpress` packages for example). Maybe it won't be the case with `jest` though.

tyxla commented on packages/scripts/utils/test/index.js:
```diff
@@ -29,26 +29,28 @@ import {
 
-	jest.spyOn( module, 'getPackagePath' );
-	jest.spyOn( module, 'hasPackageProp' );
+	( () => {
+		jest.spyOn( module, 'getPackagePath' );
+		jest.spyOn( module, 'hasPackageProp' );
+	} )();
```
Why was this IIFE necessary?

tyxla commented on test/native/jest.config.js:
```diff
@@ -13,48 +13,45 @@ if ( process.env.TEST_RN_PLATFORM ) {
 	preset: 'react-native',
-	setupFiles: [ '<rootDir>/' + configPath + '/setup.js' ],
-	setupFilesAfterEnv: [ '<rootDir>/' + configPath + '/setup-after-env.js' ],
+	setupFiles: [ '<rootDir>/setup.js' ],
+	setupFilesAfterEnv: [ '<rootDir>/setup-after-env.js' ],
```
Nice simplification here 👍 

It seems like it's arguable which one is better since now we have to specify a bunch of additional traversing below. `../../`.

jsnajdr replied:
I had to change the root, and also to introduce a special `babel.config.js` in `packages/e2e-tests`, because of https://github.com/facebook/jest/pull/12689. After that change Jest (namely `babel-jest`) will find `test/native/babel.config.js` only if the `rootDir` is `test/native`. So I had to change it, although it introduces extra `../..` traversals.

tyxla commented on test/native/jest.config.js:
```diff
@@ -13,48 +13,45 @@ if ( process.env.TEST_RN_PLATFORM ) {
-	testPathIgnorePatterns: [
-		'/node_modules/',
-		'/__device-tests__/',
-		'<rootDir>/.*/build/',
-		'<rootDir>/.*/build-module/',
```
Do we still need to keep these ignore patterns? Or is there a good reason we're removing them?

jsnajdr replied:
I'm removing them because:
1. They don't work with `../..` added, like in `<rootDir>/../../.*/build-module`. On regular expression patterns, Jest won't do the path resolution that would resolve `a/b/c/../..` into `a`, and the patterns with `../..` not removed won't match anything.
2. These ignore patterns were a workaround for the fact that there are unit tests in `packages/react-native-editor/src/parser` directory, and that directory is not excluded from build. I solved that by moving them into `packages/react-native-editor/src/test`. After that, these ignores are no longer needed.

tyxla commented on test/unit/scripts/resolver.js:
```diff
@@ -0,0 +1,35 @@
+		...options,
+		// Use packageFilter to process parsed `package.json` before the resolution (see https://www.npmjs.com/package/resolve#resolveid-opts-cb)
+		packageFilter: ( pkg ) => {
+			// This is a workaround for https://github.com/uuidjs/uuid/pull/616
+			//
+			// jest-environment-jsdom 28+ tries to use browser exports instead of default exports,
+			// but uuid only offers an ESM browser export and not a CommonJS one. Jest does not yet
+			// support ESM modules natively, so this causes a Jest error related to trying to parse
+			// "export" syntax.
+			//
+			// This workaround prevents Jest from considering uuid's module-based exports at all;
+			// it falls back to uuid's CommonJS+node "main" property.
+			//
+			// Once we're able to migrate our Jest config to ESM and a browser crypto
+			// implementation is available for the browser+ESM version of uuid to use (eg, via
+			// https://github.com/jsdom/jsdom/pull/3352 or a similar polyfill), this can go away.
```
Thanks for the thorough explanation 👍

tyxla reviewed (commented):
Thanks for the great work on this one @jsnajdr 🙌 

I think we're missing quite a few CHANGELOG entries. This is a big upgrade, breaking for some of the packages.

Also, I didn't see an obvious difference when running the tests locally, I wonder if you observed a performance improvement by using a newer `jest` with a newer `jsdom`.

SimenB commented on test/unit/scripts/resolver.js:
```diff
@@ -0,0 +1,35 @@
+		packageFilter: ( pkg ) => {
+			// This is a workaround for https://github.com/uuidjs/uuid/pull/616
+			//
+			// jest-environment-jsdom 28+ tries to use browser exports instead of default exports,
+			// but uuid only offers an ESM browser export and not a CommonJS one. Jest does not yet
+			// support ESM modules natively, so this causes a Jest error related to trying to parse
+			// "export" syntax.
```
note that uuid v9 does expose CJS entrypoint for browser env.

You might still need this custom resolver, just pointing out that the comment is somewhat stale 🙂

jsnajdr replied:
Thanks, I see the browser export was fixed in https://github.com/uuidjs/uuid/pull/642. I'll change the comment to mention "uuid v8" explicitly. After we upgrade to uuid v9, we can remove this workaround. I'd rather do the v9 upgrade in a separate PR, I'm not even sure if it will be feasible, as there are also 3rd party libraries which can use old version of `uuid`.

jsnajdr committed:
Add changelog entries and update docs

jsnajdr commented:
I did some final updates for the PR:
- when I started working on this in January, the latest Jest version was 29.3.1, today it's 29.5.0. I updated the PR to upgrade to 29.5.0. One of the benefits is that 29.5.0 fixes a hoisting bug we identified during this upgrade, and I could remove the "avoid unwanted hoisting" workarounds.
- updated peer dependencies of all relevant packages to `jest >= 29`
- added changelog entries to packages where the Jest upgrade is a breaking change

@tyxla @gziolo If the CI checks come back as green, could you do a final review of this upgrade?

gziolo reviewed (approved):
Looks good to me. I don't have the knowledge to review some of the changes related to React Native and the custom resolver that seems to be targeting ES Modules in `node_modules`. However, everything is green on CI and the part for the web looks excellent. Great work on that upgrade!

jsnajdr commented:
> the custom resolver that seems to be targeting ES Modules in `node_modules`.

This works around an issue where the new Jest tries to use the `exports` field in `package.json` and in JSDOM environment, loads the browser version of the package. But some packages export an ESM module as the browser version, instead of CJS, and Jest can't handle that.

The fixes I did were inspired by this very thoughtful comment in another project that did the Jest 29 upgrade last year: https://github.com/microsoft/accessibility-insights-web/pull/5421#issuecomment-1109168149

tyxla reviewed (commented):
Great work, thank you @jsnajdr 🙌

jsnajdr commented:
> We should also update the README to reflect it:

@gziolo The `jest-preset-default` README is being updated in #48925. It needed more updates than just the `timers`.

jsnajdr commented:
After merge, CI checks are failing because `npm ci` is failing for some reason. I'm investigating this.

gziolo commented:
> After merge, CI checks are failing because `npm ci` is failing for some reason. I'm investigating this.

Most likely nothing to worry about. It often happens after updating the lock file. I believe it's a cache issue and restarting failed jobs helps. See https://github.com/WordPress/gutenberg/issues/33424 for some similar issues in the past.

jsnajdr commented:
> Most likely nothing to worry about.

Happy to hear that 🙂 Locally, running `npm ci` succeeds, even on clean tree after `distclean`, and neither `npm ci` nor `npm i` produce any additional lockfile changes. There's nothing I could fix.

ktmn commented:
Updated my `@wordpress/scripts` from `25.2.0` to `26.3.0`, running `wp-scripts test-unit-js` no longer works.

```
Test suite failed to run

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:

    /.../my-project/node_modules/.pnpm/uuid@8.3.2/node_modules/uuid/dist/esm-browser/index.js:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){export { default as v1 } from './v1.js';
                                                                                      ^^^^^^

    SyntaxError: Unexpected token 'export'

       8 | } from '@wordpress/data'
       9 |
    > 10 | import {
         | ^
      11 |      store as editPostStore,
      12 | } from '@wordpress/edit-post'
      13 |

      at Runtime.createScriptFromCode (../../node_modules/.pnpm/jest-runtime@29.5.0/node_modules/jest-runtime/build/index.js:1495:14)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+blocks@12.9.0_react@18.2.0/node_modules/@wordpress/blocks/build/api/@wordpress/blocks/src/api/factory.js:4:1)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+blocks@12.9.0_react@18.2.0/node_modules/@wordpress/blocks/build/api/@wordpress/blocks/src/api/utils.js:20:1)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+blocks@12.9.0_react@18.2.0/node_modules/@wordpress/blocks/build/store/@wordpress/blocks/src/store/reducer.js:15:1)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+blocks@12.9.0_react@18.2.0/node_modules/@wordpress/blocks/build/store/@wordpress/blocks/src/store/index.js:9:1)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+blocks@12.9.0_react@18.2.0/node_modules/@wordpress/blocks/build/@wordpress/blocks/src/index.js:11:1)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+edit-post@7.9.0_@babel+core@7.21.5_@types+react@18.2.0_react-dom@18.2.0_react@18.2.0/node_modules/@wordpress/edit-post/build/@wordpress/edit-post/src/index.js:4:1)
      ...
```

jsnajdr commented:
Hi @ktmn :wave: You're running into an issue with `uuid`, discussed in https://github.com/WordPress/gutenberg/pull/47388#discussion_r1118596522 and https://github.com/WordPress/gutenberg/pull/47388#discussion_r1118681987, and described in this code comment:

https://github.com/WordPress/gutenberg/blob/trunk/test/unit/scripts/resolver.js#L10-L22

One way how to solve it is to upgrade `uuid` in your project to v9. But you might not be able to do that, as it's often a transitive dependency of a package you don't have control over.

If that doesn't work, we should probably include the custom Jest resolver that works around these issues in the Jest config for `wp-scripts test-unit-js`.
